elasticsearch/templates/statefulset.yaml

```diff
@@ -191,6 +191,48 @@ spec:
           - name: node.{{ $role }}
             value: "{{ $enabled }}"
           {{- end }}
+        {{- if and (eq .Values.roles.master "true") (ge (int .Values.esMajorVersion) 7) }}
+        lifecycle:
+          preStop:
+            exec:
+              command:
+              - sh
+              - -c
+              - |
+                #!/usr/bin/env bash -e
+                # Exclude node from voting config to prevent https://github.com/elastic/helm-charts/issues/63
+                # reference: https://www.elastic.co/guide/en/elasticsearch/reference/7.0/modules-discovery-adding-removing-nodes.html#modules-discovery-removing-nodes
+                http () {
+                    local path="${1}"
+                    local method="${2:-GET}"
+                    if [ -n "${ELASTIC_USERNAME}" ] && [ -n "${ELASTIC_PASSWORD}" ]; then
+                      BASIC_AUTH="-u ${ELASTIC_USERNAME}:${ELASTIC_PASSWORD}"
+                    else
+                      BASIC_AUTH=''
+                    fi
+                    curl -X${method} -s -k --fail ${BASIC_AUTH} {{ .Values.protocol }}://127.0.0.1:{{ .Values.httpPort }}${path}
+                }
+
+                NODE_NAME=$(awk 'BEGIN {print ENVIRON["node.name"]}')
+                echo "Exclude the node ${NODE_NAME} from voting config"
+
+                http "/_cluster/voting_config_exclusions/${NODE_NAME}" "POST"
+
+                while true ; do
```

Comment:
Being an infinite loop seems to be ok here because the termination of pods waits for the graceful timeout for the preStop condition to finish: https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods.